---
title: "Tools for Genomic Teaching"
author: "Tiago Olivoto"
date: "`r Sys.Date()`"
always_allow_html: yes
output:
  html_document:
    df_print: paged
fig_caption: yes
vignette: >
  %\VignetteIndexEntry{Tools for Genomic Teaching}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  rmarkdown.html_vignette.check_title = FALSE,
  fig.width = 10
)
```



# Genetic code
The function `aminoacid()` returns the aminoacid generated by the informed codon(s), following the desired translation table informed in `transl_table` argument. Available tables are presented bellow. See more at
   \url{https://www.ncbi.nlm.nih.gov/Taxonomy/taxonomyhome.html/index.cgi?chapter=cgencodes}


 1. The Standard Code (default)
 2. The Vertebrate Mitochondrial Code
 3. The Yeast Mitochondrial Code
 4. The Mold, Protozoan, and Coelenterate Mitochondrial Code and the
 Mycoplasma/Spiroplasma Code
 5. The Invertebrate Mitochondrial Code
 6. The Ciliate, Dasycladacean and Hexamita Nuclear Code
 9. The Echinoderm and Flatworm Mitochondrial Code
 10. The Euplotid Nuclear Code
 11. The Bacterial, Archaeal and Plant Plastid Code
 12. The Alternative Yeast Nuclear Code
 13. The Ascidian Mitochondrial Code
 14. The Alternative Flatworm Mitochondrial Code
 16. Chlorophycean Mitochondrial Code
 21. Trematode Mitochondrial Code
 22. Scenedesmus obliquus Mitochondrial Code
 23. Thraustochytrium Mitochondrial Code
 24. Rhabdopleuridae Mitochondrial Code
 25. Candidate Division SR1 and Gracilibacteria Code
 26. Pachysolen tannophilus Nuclear Code
 27. Karyorelict Nuclear Code
 28. Condylostoma Nuclear Code
 29. Mesodinium Nuclear Code
 30. Peritrich Nuclear Code
 31. Blastocrithidia Nuclear Code
 33. Cephalodiscidae Mitochondrial UAA-Tyr Code
 
 Use the function `gen_code()` to get the desired genetic code
```{r}
library(genomics)
gen_code()
gen_code(transl_table = 2)

```
 
# Random sequences of nucleotides
The function `rand_nuc_seq()` generate random sequences of nucleotides (i.e., A, T, C, and G) for DNA and A, U, C, and G for RNA. The random sequences are generated so that a start codon (AUG for RNA or ATG for DNA) is inserted between the first and third bases in the chain. A stop codon (UAA, UAG, or UGA for standard genetic code) is randomly inserted at the end of the chain. A message is returned with the position and stop codon used. Note that the argument `seed` ensures reproducibility.

```{r}
# a random sequence of 30 bases (DNA)
(dna <- rand_nuc_seq(30, seed = 20))

# a random sequence of 30 bases (RNA)
(rna <- rand_nuc_seq(30, type = "rna", seed = 50))

```



# Transcription
The following examples, transcripts a DNA (code) to a 'mold', i.e., the antiparallel sequence of the code (`m_dna`), and a pre RNA sequence (`m_rna`).
```{r}
dna_to_rna(dna)
```

Alternatively, it is possible to reverse the transcription of mRNA back to DNA with `rna_to_dna()`.
```{r}
rna_to_dna(rna)
```



# Translation
## Standard genetic code
The following example translates three codons considering the standard genetic code (`transl_table = 1`). Stop codons are shown was `"*"`. By default, a [one-letter code](http://130.88.97.239/bioactivity/aacodefrm.html) is used. Alternatively, one can use the argument `type` to change the output type:
* `type = "olc"` for one-letter code, e.g., "V".
* `type = "tlc"` for three-letter code, e.g., "Val".
* `type = "name"` for the complete name, e.g., "Valine".

 
```{r}
library(genomics)
codons <- c("AUG", "GUA", "ACG")
aminoacid(codons)
aminoacid(codons, type = "tlc")
aminoacid(codons, type = "name")
```

The function `translate()` can be used to translate nucleotide sequences. Let's translate a sequence of 20 nucleotides using the standard genetic code.


```{r}
# 'code' DNA
rna2 <- rand_nuc_seq(n = 20, type = "rna", seed = 25)
rna2

# a one-letter code
translate(rna2)
```


By default, translation begins only when the `start_codon` (AUG) is found. To start the translation with the first codon of the sequence, use `start_codon_first = FALSE`. Note that in the following example, the second codon (UGA) is a stop codon. 

```{r}
highlight_pattern(rna2, "AUG")

# translation begins with the first nucleotide
translate(rna2, start_codon_first = FALSE)


# one-leter code
# translation begins with the first nucleotide
# translation don't stop when reaching the stop codon
translate(rna2,
          type = "olc",
          start_codon_first = FALSE,
          no_stop = TRUE)

# a three-leter code
# translation begins with the first nucleotide
# translation don't stop when reaching the stop codon
translate(rna2,
          type = "name",
          start_codon_first = FALSE,
          no_stop = TRUE,
          sep = "-")
```


Check the translation at https://web.expasy.org/translate/


## Alternative translation tables
Using the argument `transl_table` it is possible to change the default standard code table. In the following example, the *The Bacterial, Archaeal and Plant Plastid Code* (translation table 11 is used).

```{r}
seq <- rand_nuc_seq(200, type = "rna", seed = 5)

translate(seq, transl_table = 11)
```

