#' Tools for the genetic code
#'
#' * `aminoacid()` returns the aminoacid generated by the informed codon(s).
#' * `gen_code()` returns all the amino acids considering the translation table
#' used.
#'
#' @details The translation is performed according to the desired translation
#'   table. See more at
#'   \url{https://www.ncbi.nlm.nih.gov/Taxonomy/taxonomyhome.html/index.cgi?chapter=cgencodes}
#'
#'
#' 1. The Standard Code (default)
#' 2. The Vertebrate Mitochondrial Code
#' 3. The Yeast Mitochondrial Code
#' 4. The Mold, Protozoan, and Coelenterate Mitochondrial Code and the
#' Mycoplasma/Spiroplasma Code
#' 5. The Invertebrate Mitochondrial Code
#' 6. The Ciliate, Dasycladacean and Hexamita Nuclear Code
#' 9. The Echinoderm and Flatworm Mitochondrial Code
#' 10. The Euplotid Nuclear Code
#' 11. The Bacterial, Archaeal and Plant Plastid Code
#' 12. The Alternative Yeast Nuclear Code
#' 13. The Ascidian Mitochondrial Code
#' 14. The Alternative Flatworm Mitochondrial Code
#' 16. Chlorophycean Mitochondrial Code
#' 21. Trematode Mitochondrial Code
#' 22. Scenedesmus obliquus Mitochondrial Code
#' 23. Thraustochytrium Mitochondrial Code
#' 24. Rhabdopleuridae Mitochondrial Code
#' 25. Candidate Division SR1 and Gracilibacteria Code
#' 26. Pachysolen tannophilus Nuclear Code
#' 27. Karyorelict Nuclear Code
#' 28. Condylostoma Nuclear Code
#' 29. Mesodinium Nuclear Code
#' 30. Peritrich Nuclear Code
#' 31. Blastocrithidia Nuclear Code
#' 33. Cephalodiscidae Mitochondrial UAA-Tyr Code
#'
#'
#' @param codon A character vector (upper- or lower-case letters) with the 5'-3'
#'   codon (e.g., "AUG").
#' @param transl_table The translation table used. Defaults to 1 (standard
#'   genetic code). See more in `Details` section.
#' @param type The type of output. One of
#' * `"olc"` for one-letter code, e.g., "V".
#' * `"tlc"` for three-letter code, e.g., "Val".
#' * `"name"` for the complete name, e.g., "Valine".
#' @return
#' * `aminoacid()` returns a character vector in the format `code_letter_name` of the given aminoacid(s)
#' * `gen_code()` returns a data.frame object with the following columns.
#' * `fb` The first base of the codon.
#' * `sb` The second base of the codon.
#' * `tb` The third base of the codon.
#' * `codon` The codon (5'-3').
#' * `olc` The one letter code.
#' * `tlc` The three letter code.
#' * `name` The name of the aminoacid.
#'
#' The stop codons are depicted by `"*"`
#' @export
#' @name gen_code
#' @import poorman
#' @importFrom  stringr str_replace_all str_split str_locate str_locate_all str_sub str_detect str_c str_sub<-
#' @importFrom utils read.csv
#'
#' @examples
#' aac <- c("AUG", "CGA", "UGA")
#' aminoacid(aac)
#' aminoacid(aac, transl_table = 2)
#'
#' # complete genetic code
#' gen_code()
#'

aminoacid <- function(codon, transl_table = 1, type = "olc"){
  path <- system.file(package = "genomics")
  gcod <- read.csv(file =  paste0(path, "/gen_code.csv"), header = T, sep = ";")
  amino <- read.csv(file =  paste0(path, "/aminoacids.csv"), header = T, sep = ";")
  if(length(codon) > 1){
    as.character(do.call(rbind, lapply(codon, aminoacid, transl_table,  type)))
  } else{
    if(!type %in% c("olc", "tlc", "name")){
      stop("Type must be one 'olc', 'tlc', or 'name'.")
    }
    codon <- tidy_sequence(codon)
    check_base(codon)
    if(nchar(codon) != 3){
      stop("Codon must be a length of 3.", call. = FALSE)
    }
    olc <- gcod[which(codon == gcod$codon), which(colnames(gcod) == paste0("tt", transl_table))]
    amino[which(amino$olc == olc), type]
  }
}

#' @export
#' @name gen_code
gen_code <- function(transl_table = 1){
  bases <- c("A", "U", "G", "C")
  gcode <-
    expand.grid(fb = bases, sb = bases, tb = bases) %>%
    arrange(fb, sb, tb) %>%
    mutate(
      codon = paste0(fb, sb, tb),
      olc = aminoacid(codon, transl_table, type = "olc"),
      tlc = aminoacid(codon, transl_table, type = "tlc"),
      name = aminoacid(codon, transl_table, type = "name")
    )
  rownames(gcode) <- NULL
  return(gcode)
}
